app/file.go

```diff
@@ -418,23 +424,41 @@ func (a *App) DoUploadFile(now time.Time, rawTeamId string, rawChannelId string,
 		// Check dimensions before loading the whole thing into memory later on
 		if info.Width*info.Height > MaxImageSize {
 			err := model.NewAppError("uploadFile", "api.file.upload_file.large_image.app_error", map[string]interface{}{"Filename": filename}, "", http.StatusBadRequest)
-			return nil, err
+			return nil, data, err
 		}
 
 		nameWithoutExtension := filename[:strings.LastIndex(filename, ".")]
 		info.PreviewPath = pathPrefix + nameWithoutExtension + "_preview.jpg"
 		info.ThumbnailPath = pathPrefix + nameWithoutExtension + "_thumb.jpg"
 	}
 
+	if a.PluginsReady() {
+		pluginContext := &plugin.Context{}
+		var rejectionReason string
+		a.Plugins.RunMultiPluginHook(func(hooks plugin.Hooks) bool {
+			var newBytes bytes.Buffer
+			info, rejectionReason = hooks.FileWillBeUploaded(pluginContext, info, bytes.NewReader(data), &newBytes)
+			rejected := info == nil
+			if !rejected && newBytes.Len() != 0 {
```

Comment:
Why the `newBytes.Len() != 0` check? Can the plugin not zero out the bytes but still allow the upload?

Reply:
This is an optimization so that in the case where the plugin doesn't want to modify the file it doesn't have to write the whole thing back to the server. Returning nothing is the signal for not modified.

Your right it precludes the case where the plugin wants a 0 byte file uploaded. I just don't see that as useful.

Reply:
Definitely agree with the optimization, and that a 0 byte file upload isn't useful. 

Returning "nothing" without a "reason" seems like a nice pattern for, "allow without modification". I'm wondering if we should do the same for `MessageWillBePosted`, where `return nil, ""` has similar semantics. This avoids the shoot-yourself-in-the-foot behaviour we discussed before.